File: src/supportScripts/connectorWorkFreeDays/connectorWorkFreeDays.py
import requests
import json
from datetime import datetime
from datetime import date

class WorkFreeDays:
    
    def __init__(self, iso2):
        
        self.HOST = "https://date.nager.at/api/v3/PublicHolidays";
        
        self.ISO2 = iso2
        
        print("WorkFreeDays")
    
    # TODO: Save API responses and optimize code
    def getPublicHolidays(self, isoDate):
        
        isPublicHoliday = False;
        
        URL = "%s/%d/%s" % (self.HOST, isoDate.year, self.ISO2)
        
        print(URL);
        
        response = requests.request("GET", URL)
        
        print(response.text)
        
        print(isoDate.isoformat())
        
        for day in json.loads(response.text):
            if(day["date"] == isoDate.isoformat()):
                isPublicHoliday = True;
                
        return isPublicHoliday;
        
    def getDayInfo(self, DATE):
        
        isoDate = date.fromisoformat(DATE)
        
        dayInfo = { "WorkFreeDay": False }
        
        print(isoDate.month)
        
        if(isoDate.weekday() == 5): dayInfo["WorkFreeDay"] = True; # Saturday
        if(isoDate.weekday() == 6): dayInfo["WorkFreeDay"] = True; # Sunday
        
        if(dayInfo["WorkFreeDay"] == False):
            dayInfo["WorkFreeDay"] = self.getPublicHolidays(isoDate)
            
        return dayInfo;

def TEST():
    
    countryIso2 = "SI";
    
    API = WorkFreeDays(countryIso2)
    
    print(API.getDayInfo("2024-10-08")) # Normal working day
    print(API.getDayInfo("2024-11-01")) # Public holiday - 1st november
    print(API.getDayInfo("2024-10-12")) # Weekend - saturday
    
    # TODO: Unit tests
    
    
TEST()

File: src/supportScripts/connectorSIPX/connectorWorkFreeDays.js
let axios = require("axios");
let Moment = require('moment');
var sprintf = require('sprintf-js').sprintf;

class connectorWorkFreeDays
{
    constructor()
    {
        this.freeDays = [];
    }
    
    async setPublicHolidays(year, iso2)
    {
        let publicHolidays = this.freeDays.find(x => x.iso2);
        
        if(!publicHolidays)
        {
            publicHolidays = { iso2: iso2, dates: [] };
            
            this.freeDays.push(publicHolidays);
        }
        
        let config =
        {
            method: 'get',
            url: sprintf("%s/%d/%s", "https://date.nager.at/api/v3/PublicHolidays", year, iso2)
        };
       
        return await axios.request(config).then((response) =>
        {
            publicHolidays.dates =
            [
                ...publicHolidays.dates,
                ...response.data.filter(x => x.types.includes("Public")).map(x => { return { date: Moment(x.date), name: x.name }; })
            ];
        })
        .catch((error) =>
        {
            console.log(error);
            
            throw new Error("setPublicHolidays");
        });
    }
    
    async isWorkFreeDay(date)
    {
        if(Moment(date).day() === 6) { return true; } // Saturday
        if(Moment(date).day() === 7) { return true; } // Sunday
        
        if(this.freeDays.includes(Moment(date)))
        {
            return true;
        }
        
        return false;
    }
}

async function INIT()
{
    let API = new connectorWorkFreeDays();
        
    await API.setPublicHolidays(2020, 'SI');
    await API.setPublicHolidays(2021, 'SI');
    await API.setPublicHolidays(2022, 'SI');
    await API.setPublicHolidays(2023, 'SI');
    await API.setPublicHolidays(2024, 'SI');
    await API.setPublicHolidays(2025, 'SI');
    
    return API;
}

async function TEST()
{
    if(require.main === module)
    {
        console.log("Connector [TEST]: WorkFreeDays");
        
        let API = await INIT();
        
        console.log(API.isWorkFreeDay("2025-02-01"));
    }
}

TEST();

module.exports = INIT;
